r-package/grf/R/local_linear_causal_tuning.R

```diff
@@ -0,0 +1,68 @@
+#' Local linear forest tuning
+#' 
+#' Finds the optimal ridge penalty for local linear causal prediction.
+#'
+#' @param forest The forest used for prediction.
+#' @param linear.correction.variables Variables to use for local linear prediction. If left null,
+#'          all variables are used.
+#' @param ll.weight.penalty Option to standardize ridge penalty by covariance (TRUE),
+#'                            or penalize all covariates equally (FALSE). Defaults to FALSE.
+#' @param num.threads Number of threads used in training. If set to NULL, the software
+#'                    automatically selects an appropriate amount.
+#' @param lambda.path Optional list of lambdas to use for cross-validation.
+#' @return A list of lambdas tried, corresponding errors, and optimal ridge penalty lambda.
+#'
+#' @examples \dontrun{
+#' # Find the optimal tuning parameters.
+#' n = 50; p = 10
+#' X = matrix(rnorm(n*p), n, p)
+#' W = rbinom(n, 1, 0.5)
+#' Y = pmax(X[,1], 0) * W + X[,2] + pmin(X[,3], 0) + rnorm(n)
+#'
+#' forest = causal_forest(X,Y,W)
+#' tuned.lambda = tune_ll_causal_forest(forest)
+#'
+#' # Use this parameter to predict from a local linear causal forest.
+#' predictions = predict(forest, linear.correction.variables = 1:p, lambda = tuned.lambda)
+#' }
+#'
+#' @export
+tune_ll_causal_forest <- function(forest,
+                                  linear.correction.variables = NULL,
+                                  ll.weight.penalty = FALSE,
+                                  num.threads = NULL,
+                                  lambda.path = NULL) {
+  forest.short = forest[-which(names(forest) == "X.orig")]
+
+  X = forest[["X.orig"]]
+  Y = forest[["Y.orig"]]
+  W = forest[["W.orig"]]
+  data = create_data_matrices(X, Y, W)
+  outcome.index = ncol(X) + 1
+  treatment.index = ncol(X) + 2
+  instrument.index = treatment.index
+
+  # Validate variables
+  num.threads = validate_num_threads(num.threads)
+  linear.correction.variables = validate_ll_vars(linear.correction.variables, ncol(X))
+  lambda.path = validate_ll_path(lambda.path)
+
+  # Subtract 1 to account for C++ indexing
+  linear.correction.variables = linear.correction.variables - 1
+
+  # Find sequence of predictions by lambda
+  prediction.object = ll_causal_predict_oob(forest.short, data$default, data$sparse, outcome.index,
+      treatment.index, instrument.index, lambda.path, ll.weight.penalty, linear.correction.variables, num.threads)
+  predictions = prediction.object$predictions
+
+  # Use R-learner to estimate tau
+  fit = rlearner::rboost(X, W, Y)
```

Comment:
I don't think we need a dependency on the `rlearner` package -- could we just compute the r-learner loss here directly?

Reply:
Yes agreed, there must have been a miscommunication here. The idea is to minimize the R-loss \hat{L} defined in (6) from https://arxiv.org/pdf/1712.04912.pdf to choose lambda. The nuisance components \hat{e} and \hat{m} used to form this loss function are the same as those used to grow the forest in the first place -- there's no need to train any new predictor here.

Reply:
Ohh good point- this is doing lots of extra work. Replaced, thanks for the find!